ios/Classes/AblyStreamsChannel.m

```diff
@@ -0,0 +1,138 @@
+  self = [super init];
+  NSAssert(self, @"Super init cannot be nil");
+  _name = name;
+  _messenger = messenger;
+  _codec = codec;
+  return self;
+}
+
+- (void)setStreamHandlerFactory:(NSObject<FlutterStreamHandler> *(^)(id))factory {
+  if (!factory) {
+    [_messenger setMessageHandlerOnChannel:_name binaryMessageHandler:nil];
+    return;
+  }
+  
+  _streams = [NSMutableDictionary new];
+  _listenerArguments = [NSMutableDictionary new];
+  FlutterBinaryMessageHandler messageHandler = ^(NSData* message, FlutterBinaryReply callback) {
+    FlutterMethodCall* call = [self->_codec decodeMethodCall:message];
+    NSArray *methodParts = [call.method componentsSeparatedByString:@"#"];
+    
+    if (methodParts.count != 2) {
+      callback(nil);
+      return;
+    }
+    
+    NSInteger keyValue = [methodParts.lastObject integerValue];
+    if(keyValue == 0) {
+      callback([self->_codec encodeErrorEnvelope:[FlutterError errorWithCode:@"error" message:[NSString stringWithFormat:@"Invalid method name: %@", call.method] details:nil]]);
+      return;
+    }
+    
+    NSNumber *key = [NSNumber numberWithInteger:keyValue];
+    
+    if ([methodParts.firstObject isEqualToString:@"listen"]) {
+        [self listenForCall:call withKey:key usingCallback:callback andFactory:factory];
+    } else if ([methodParts.firstObject isEqualToString:@"cancel"]) {
+        [self cancelForCall:call withKey:key usingCallback:callback andFactory:factory];
+    } else {
+      callback(nil);
+    }
+  };
+  
+  [_messenger setMessageHandlerOnChannel:_name binaryMessageHandler:messageHandler];
+}
+
+- (void) reset{
+    for (NSString* key in self->_streams) {
+        AblyStreamsChannelStream *stream = self->_streams[key];
+        [stream.handler onCancelWithArguments:nil];
+        [_streams removeObjectForKey:key];
+        [_listenerArguments removeObjectForKey:key];
+    }
+}
+
+- (void)listenForCall:(FlutterMethodCall*)call withKey:(NSNumber*)key usingCallback:(FlutterBinaryReply)callback andFactory:(NSObject<FlutterStreamHandler> *(^)(id))factory {
+  AblyStreamsChannelStream *stream = [AblyStreamsChannelStream new];
+  stream.sink = ^(id event) {
+    NSString *name = [NSString stringWithFormat:@"%@#%@", self->_name, key];
+    
+    if (event == FlutterEndOfEventStream) {
+      [self->_messenger sendOnChannel:name message:nil];
+    } else if ([event isKindOfClass:[FlutterError class]]) {
+      [self->_messenger sendOnChannel:name
+                              message:[self->_codec encodeErrorEnvelope:(FlutterError*)event]];
+    } else {
+      [self->_messenger sendOnChannel:name message:[self->_codec encodeSuccessEnvelope:event]];
+    }
+  };
+  stream.handler = factory(call.arguments);
+  
+  [_streams setObject:stream forKey:key];
+  [_listenerArguments setObject:call.arguments forKey:key];
+  
+  FlutterError* error = [stream.handler onListenWithArguments:call.arguments eventSink:stream.sink];
+  if (error) {
```

Comment:
I notice this logic is used in a couple of places. Could you move this to the implementation of `encodeErrorEnvelope:` to decide? It could accept `nil` values for `error` and return a success envelope in that case. This this line would just become:

```suggestion
  callback([self->_codec encodeErrorEnvelope:error]);
```

With corresponding clear commentary on that interface to explain this behaviour.

Reply:
Bad news! `encodeErrorEnvelope` method is from flutter's source.

Reply:
ok, perhaps an opportunity for your own wrapper - even if just a static local function?